src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/PackageRestoreUnconfiguredDataSource.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            // Always use the first TFM listed in project to provide consistent behavior
+            ProjectRestoreUpdate update = updates.First();
+            string propertyValue = propertyGetter(update.RestoreInfo);
+
+            // Every config should had same value
+            bool hasConflicts = updates.Select(u => propertyGetter(u.RestoreInfo))
+                                       .Distinct(StringComparers.PropertyNames)
+                                       .Count() > 1;
+
+            if (hasConflicts)
+            {
+                ReportDataSourceUserFault(new Exception(string.Format(CultureInfo.CurrentCulture, VSResources.Restore_PropertyWithInconsistentValues, propertyName, propertyValue, update.ProjectConfiguration)),
+                                          ProjectFaultSeverity.LimitedFunctionality,
+                                          ContainingProject);
+            }
+
+            return propertyValue;
+        }
+
+        private IVsReferenceItems ResolveToolReferenceConflicts(IEnumerable<ProjectRestoreUpdate> updates)
+        {
+            var references = new Dictionary<string, IVsReferenceItem>(StringComparers.ItemNames);
+
+            foreach (ProjectRestoreUpdate update in updates)
+            {
+                foreach (IVsReferenceItem reference in update.RestoreInfo.ToolReferences)
+                {
+                    if (references.TryGetValue(reference.Name, out IVsReferenceItem existingReference))
+                    {
+                        ReportUserFaultIfToolReferenceConflict(existingReference, reference);
+                        continue;
+                    }
+
+                    references.Add(reference.Name, reference);
+                }
+            }
+
+            return new ReferenceItems(references.Values);
+        }
+
+        private void ReportUserFaultIfToolReferenceConflict(IVsReferenceItem existingReference, IVsReferenceItem reference)
+        {
+            // CLI tool references are project-wide, so if they have conflicts in names, 
+            // they must have the same metadata, which avoids from having to condition 
+            // them so that they only appear in one TFM.
+            if (!ReferenceItemEqualityComparer.Instance.Equals(existingReference, reference))
+            {
+                ReportDataSourceUserFault(new Exception(string.Format(CultureInfo.CurrentCulture, VSResources.Restore_DuplicateToolReferenceItems, existingReference.Name)),
+                                          ProjectFaultSeverity.LimitedFunctionality,
+                                          ContainingProject);
+            }
+        }
+
+        private static IVsTargetFrameworks GetAllTargetFrameworks(IEnumerable<ProjectRestoreUpdate> updates)
+        {
+            IEnumerable<IVsTargetFrameworkInfo> targetFrameworks = updates.SelectMany(t => t.RestoreInfo.TargetFrameworks.Cast<IVsTargetFrameworkInfo>());
```

Comment:
do we want to retain a LinQ delegation inside our data structure? this can produce more complex in memory structure. and you rerun the queue every time the property is used. maybe getting the result out can be a simpler solution?

Reply:
The LINQ query is realized in the the TargetFrameworks constructor, it doesn't outlive this method.